packages/language-service/lib/utils.js

```diff
@@ -0,0 +1,270 @@
+
+const componentStart = `
+/**
+ * Render the MDX contents.
+ *
+ * @param {MDXContentProps} props The props that have been passed to the MDX component.
+ */
+export default function MDXContent(props) {
+  return <>
+`
+const componentEnd = `
+  </>
+}
+
+// @ts-ignore
+/** @typedef {Props} MDXContentProps */
+`
+
+const fallback = 'export {}\n'
+
+const whitespaceRegex = /\s/u
+
+/**
+ * @param {OffsetRange[]} positions
+ * @param {number} index
+ * @returns {boolean} XXX
+ */
+function shouldShow(positions, index) {
+  return positions.some(([start, end]) => start <= index && index < end)
+}
+
+/**
+ * @param {OffsetRange} a
+ * @param {OffsetRange} b
+ * @returns {number} XXX
+ */
+function compareRanges(a, b) {
+  return a[0] - b[0] || a[1] - b[1]
+}
+
+/**
+ * @param {Parent} node
+ * @returns {number | undefined} XXX
+ */
+function findFirstOffset(node) {
+  for (const child of node.children) {
+    const start = child.position?.start?.offset
+    if (start !== undefined) {
+      return start
+    }
+  }
+}
+
+/**
+ * @param {Parent} node
+ * @returns {number | undefined} XXX
+ */
+function findLastOffset(node) {
+  for (let index = node.children.length - 1; index >= 0; index--) {
+    const end = node.children[index].position?.end?.offset
+    if (end !== undefined) {
+      return end
+    }
+  }
+}
+
+/**
+ * Convert MDX into JavaScript with JSX.
+ *
+ * MDX can be categorized in 3 types of content:
+ *
+ * 1. Markdown content; This is not relevant when offering TypeScript based intellisense.
+ * 2. ESM content; This includes JavaScript imports and exports. When MDX is compiled, this content
+ *    is moved to the top-level scope.
+ * 3. JSX content; This includes JSX elements and expressions. When MDX is compiled, this content is
+ *    moved into a function named `MDXContent`.
+ *
+ * The problem is that ESM and JSX can be mixed inside MDX, so the function body of `MDXContent` can
+ * be mixed with content in the top-level scope. To turn MDX into JavaScript that the TypeScript
+ * language service understands, we split the ESM and JSX content, by creating a copy of the
+ * original document and replacing any character that doesn’t fall into that category with
+ * whitespace.
+ *
+ * The JSX part is then wrapped inside an `MDXContent` function declaration and a JSX fragment. The
+ * ESM and wrapped JSX parts are concatenated to produce valid JavaScript source code which
+ * represents the JavaScript parts of MDX. This result can then be processed by the TypeScript
+ * language service. Any positional information returned by TypeScript that represents an MDX file,
+ * needs to be mapped using {@link toOriginalPosition}
+ *
+ * @see https://code.visualstudio.com/api/language-extensions/embedded-languages#language-services-sample
+ * @param {string} mdx
+ * @param {Processor} processor
+ * @returns {MDXSnapshot} JavaScript code that matches the MDX code, but shadowed.
+ */
+export function mdxToJsx(mdx, processor) {
```

Comment:
Some good candidates to externalize as utilities here

Reply:
I think it’s too specific to this use case TBH. It will also need updates for future features, e.g. detecting which components are supported in the `components` prop.